contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+        Timestamp::from_seconds(self.start_time + VESTING_PERIOD)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct PeriodicVestingAccount {
+    address: Addr,
+    start_time: Timestamp,
+    periods: Vec<VestingPeriod>,
+    coin: Coin,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct DelegationData {
+    mix_identity: IdentityKey,
+    amount: Uint128,
+    block_time: Timestamp,
+}
+
+impl PeriodicVestingAccount {
+    pub fn new(
+        address: Addr,
+        coin: Coin,
+        start_time: Timestamp,
+        periods: Vec<VestingPeriod>,
+        storage: &mut dyn Storage,
+    ) -> Result<Self, ContractError> {
+        let amount = coin.amount;
+        let account = PeriodicVestingAccount {
+            address,
+            start_time,
+            periods,
+            coin,
+        };
+        set_account(storage, account.clone())?;
+        set_account_balance(storage, &account.address, amount)?;
+        Ok(account)
+    }
+
+    pub fn address(&self) -> Addr {
+        self.address.clone()
+    }
+
+    pub fn tokens_per_period(&self) -> u128 {
+        // Remainder tokens will be lumped into the last period.
+        self.coin.amount.u128() / NUM_VESTING_PERIODS as u128
+    }
+
+    fn get_next_vesting_period(&self, block_time: Timestamp) -> usize {
+        // Returns the index of the next vesting period. Unless the current time is somehow in the past or vesting has not started yet.
+        // In case vesting is over it will always return NUM_VESTING_PERIODS.
+        match self
+            .periods
+            .iter()
+            .map(|period| period.start_time)
+            .collect::<Vec<u64>>()
+            .binary_search(&block_time.seconds())
+        {
+            Ok(u) => u,
+            Err(u) => u,
+        }
+    }
+
+    fn get_current_vesting_period(&self, block_time: Timestamp) -> usize {
+        if self.get_next_vesting_period(block_time) > 0 {
+            self.get_next_vesting_period(block_time) - 1
+        } else {
+            0
+        }
+    }
+
+    fn get_delegated_with_op(
+        &self,
+        op: &dyn Fn(u64, u64) -> bool,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        let block_time = block_time.unwrap_or(env.block.time);
+        let period = self.get_current_vesting_period(block_time);
+        if period == 0 {
+            return Coin {
+                amount: Uint128(0),
+                denom: DENOM.to_string(),
+            };
+        }
+
+        let end_time = if period >= NUM_VESTING_PERIODS as usize {
+            u64::MAX
+        } else {
+            self.periods[period].end_time().seconds()
+        };
+
+        if let Some(delegations) = get_account_delegations(storage, &self.address) {
+            delegations
+                .iter()
+                .filter(|d| op(d.block_time.seconds(), end_time))
```

Comment:
I don't fully understand the purpose of passing the `lt` / `ge` op here. Couldn't we put it directly in the code here instead?

Reply:
The thing is that the function logic is identical except for the comparison operator, so this puts all that code in one place, so maintenance should be simpler